tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+        then_controls: Vec<Control>,
+        else_controls: Vec<Control>,
+    ) -> Self {
+        Control {
+            attributes: Attributes::new(),
+            value: ControlValue::If(port, group, then_controls, else_controls),
+        }
+    }
+}
+
+impl AttributeProvider for Control {
+    fn add_attribute(&mut self, attribute: Attribute) {
+        self.attributes.push(attribute);
+    }
+}
+
+impl CalyxWriter for Control {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        self.attributes.write(f)?;
+        match &self.value {
+            ControlValue::Empty => {}
+            ControlValue::Enable(group) => {
+                write!(f, "{};", group.borrow().name)?;
+            }
+            ControlValue::Seq(body) => {
+                writeln!(f, "seq {{")?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::Par(body) => {
+                writeln!(f, "par {{")?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::While(cond, group, body) => {
+                writeln!(
+                    f,
+                    "while {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::If(cond, group, body_true, body_false) => {
+                writeln!(
+                    f,
+                    "if {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body_true {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                writeln!(f, "}} else {{")?;
+                f.push();
+                for node in body_false {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// See `calyx_ir::Component`.
+pub struct Component {
+    attributes: Attributes,
+    is_comb: bool,
+    name: String,
+    inputs: Vec<Port>,
+    outputs: Vec<Port>,
+    cells: Vec<RRC<Cell>>,
+    groups: Vec<RRC<Group>>,
+    control: Control,
```

Comment:
nit: technically this should be optional since components can be fully
structural, but given the use case for this tool it's probably fine to have it
like this

Reply:
It gets defaulted to `Control::empty()`, so it'll appear in the output as `control {}`, which I think is fine for purely structural components?